akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/HostConnectionPool.scala

```diff
@@ -0,0 +1,335 @@
+        ) {
+          // FIXME: is this safe? I.e. always pulled?
+          def pushRequest(request: HttpRequest): Unit = {
+            val newRequest =
+              request.entity match {
+                case _: HttpEntity.Strict ⇒
+                  slot.onRequestEntityCompleted()
+                  request
+                case _ ⇒
+                  val (newEntity, entityComplete) = HttpEntity.captureTermination(request.entity)
+                  entityComplete.onComplete(safely {
+                    case Success(_)     ⇒ slot.onRequestEntityCompleted()
+                    case Failure(cause) ⇒ slot.onRequestEntityFailed(cause)
+                  })(ExecutionContexts.sameThreadExecutionContext)
+
+                  request.withEntity(newEntity)
+              }
+
+            requestOut.push(newRequest)
+          }
+          def close(): Unit = {
+            requestOut.complete()
+            responseIn.cancel()
+          }
+          def isClosed: Boolean = requestOut.isClosed || responseIn.isClosed
+        }
+        def openConnection(slot: Slot): SlotConnection = {
+          val requestOut = new SubSourceOutlet[HttpRequest]("PoolSlot.requestOut")
+          requestOut.setHandler(new OutHandler {
+            def onPull(): Unit = ()
+
+            override def onDownstreamFinish(): Unit = {
+              slot.debug("Connection cancelled")
+            }
+          })
+
+          val responseIn = new SubSinkInlet[HttpResponse]("PoolSlot.responseIn")
+          responseIn.setHandler(new InHandler {
+            def onPush(): Unit = {
+              val response = responseIn.grab()
+
+              slot.debug("Received response") // FIXME: add abbreviated info
+
+              response.entity match {
+                case _: HttpEntity.Strict ⇒
+                  slot.onResponseReceived(response)
+                  slot.onResponseEntityCompleted()
+                case _ ⇒
+                  val (newEntity, entityComplete) = HttpEntity.captureTermination(response.entity)
+                  entityComplete.onComplete(safely {
+                    case Success(_)     ⇒ slot.onResponseEntityCompleted()
+                    case Failure(cause) ⇒ slot.onResponseEntityFailed(cause)
+                  })(ExecutionContexts.sameThreadExecutionContext)
+
+                  slot.onResponseReceived(response.withEntity(newEntity))
+              }
+
+              if (!responseIn.isClosed) responseIn.pull()
+            }
+
+            override def onUpstreamFinish(): Unit = {
+              slot.debug("Connection completed")
+              if (slot.isConnected)
+                slot.onConnectionCompleted()
+            }
+            override def onUpstreamFailure(ex: Throwable): Unit = {
+              slot.debug("Connection failed")
+              if (slot.isConnected)
+                slot.onConnectionFailed(ex)
+            }
+          })
+          responseIn.pull()
+
+          slot.debug("Establishing connection")
```

Comment:
add info where to?

Reply:
I think right now we don't have that information here (it's hidden in the `connectionFlow` instance) but the logger from `PoolInterfaceActor` will add that information.